atomic_reactor/core.py

```diff
@@ -272,10 +282,57 @@ def __init__(self, base_url=None, retry_times=DOCKER_MAX_RETRIES,
 
         if hasattr(docker, 'AutoVersionClient'):
             client_kwargs['version'] = 'auto'
-        client_kwargs['retry'] = retry_times
+        self.retry_times = retry_times
+        client_kwargs['retry'] = self.retry_times
 
         self.d = WrappedDocker(**client_kwargs)
 
+    def retry_generator(self, function, *args, **kwargs):
+        retry_times = int(kwargs.pop('retry_times', self.retry_times))
+        retry_delay = DOCKER_BACKOFF_FACTOR
+        retry_client_statuses = DOCKER_CLIENT_STATUS_RETRY
+
+        for counter in range(retry_times + 1):
+            exc = None
+            try:
+                logs_gen = function(*args, **kwargs)
+                cmd_result = atomic_reactor.util.wait_for_command(logs_gen)
+
+            except ProtocolError as e:
+                if counter != retry_times:
```

Comment:
There's quite a bit of repetition in this method. Can we take the approach of only checking `counter != retry_times` only once. Something like this:

```python
exc = None
context = None

try:
  ...
except Protocol as e:
  exc = e
  context = e.args
exception APIError as e:
  exc = e
  context = e.reponse.content

if exc or cmd_result.is_failed():
  if counter == retry_times:
    raise RetryGeneratorException("Failed to %s image %s: %r" %" % 
                                  (function.__name__, args, context),
                                  exc)
  logger.info("retrying %s - %s on %s", function.__name__, args, context)
  time.sleep(retry_delay * (2 ** counter))

# Also handle return cases.
```

Reply:
ok reworked